external-crates/move/crates/move-core-types/src/annotated_visitor.rs

```diff
@@ -236,28 +361,91 @@ pub enum Error {
 
     #[error("invalid variant tag: {0}")]
     UnexpectedVariantTag(usize),
+
+    #[error("no layout available for value")]
+    NoValueLayout,
 }
 
 /// The null traversal implements `Traversal` and `Visitor` but without doing anything (does not
 /// return a value, and does not modify any state). This is useful for skipping over parts of the
 /// value structure.
 pub struct NullTraversal;
 
-impl Traversal for NullTraversal {
+impl<'b, 'l> Traversal<'b, 'l> for NullTraversal {
     type Error = Error;
 }
 
-#[allow(clippy::len_without_is_empty)]
-impl<'r, 'b, 'l> VecDriver<'r, 'b, 'l> {
-    fn new(bytes: &'r mut &'b [u8], layout: &'l MoveTypeLayout, len: u64) -> Self {
+impl<'c, 'b, 'l> ValueDriver<'c, 'b, 'l> {
+    pub(crate) fn new(bytes: &'c mut Cursor<&'b [u8]>, layout: Option<&'l MoveTypeLayout>) -> Self {
+        let start = bytes.position() as usize;
         Self {
             bytes,
             layout,
+            start,
+        }
+    }
+
+    /// The offset at which the value being visited starts in the byte stream.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// The current position in the byte stream.
+    pub fn position(&self) -> usize {
+        self.bytes.position() as usize
+    }
+
+    /// All the bytes in the byte stream (including the ones that have been read).
+    pub fn bytes(&self) -> &'b [u8] {
```

Comment:
This is recoverable from the current API, but I'm thinking it would also be useful to have a `peek` that returns the remaining bytes excluding the ones that have been read as well.

Reply:
Seems reasonable! There's an experimental API: `Cursor::remaining_slice` that I can borrow the nomenclature from (`remaining_bytes`).